File: python/aoc02/aoc02.py
#!/usr/bin/env python3

with open("aoc02.dat") as f:
    lines = f.read().splitlines()

# part 1

position = [0, 0]
for line in lines:
    direction, distance = line.split()
    if direction == "forward":
        position[0] += int(distance)
    elif direction == "up":
        position[1] -= int(distance)
    elif direction == "down":
        position[1] += int(distance)
    else:
        raise Exception(f"unknown direction: {direction}")

print("part 1")
print(f"position: {position}")
print(f"product: {position[0] * position[1]}")

# part 2

position = [0, 0]
aim = 0
for line in lines:
    direction, distance = line.split()
    if direction == "forward":
        position[0] += int(distance)
        position[1] += int(distance) * aim
    elif direction == "up":
        aim -= int(distance)
    elif direction == "down":
        aim += int(distance)
    else:
        raise Exception(f"unknown direction: {direction}")

print("part 2")
print(f"position: {position}")
print(f"product: {position[0] * position[1]}")

File: golang/aoc02/aoc02.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func readInput(fileName string) []string {
	file, err := os.ReadFile(fileName)
	if err != nil {
		panic(err)
	}
	lines := strings.Split(string(file), "\n")
	return lines
}

func computePositionPart1(lines []string) ([]int, int) {
	coords := []int{0, 0}
	for _, line := range lines {
		s := strings.Split(line, " ")
		direction := s[0]
		distance, err := strconv.Atoi(s[1])
		if err != nil {
			panic(err)
		}
		switch direction {
		case "forward":
			coords[0] += distance
		case "up":
			coords[1] -= distance
		case "down":
			coords[1] += distance
		}
	}
	return coords, coords[0] * coords[1]
}

func computePositionPart2(lines []string) ([]int, int) {
	coords := []int{0, 0}
	aim := 0
	for _, line := range lines {
		s := strings.Split(line, " ")
		direction := s[0]
		distance, err := strconv.Atoi(s[1])
		if err != nil {
			panic(err)
		}
		switch direction {
		case "forward":
			coords[0] += distance
			coords[1] += distance * aim
		case "up":
			aim -= distance
		case "down":
			aim += distance
		}
	}
	return coords, coords[0] * coords[1]
}

func main() {
	lines := readInput("aoc02.dat")
	var coords, product = computePositionPart1(lines)
	fmt.Printf("part 1\nposition: [%v, %v]\nproduct: %v\n", coords[0], coords[1], product)
	coords, product = computePositionPart2(lines)
	fmt.Printf("part 2\nposition: [%v, %v]\nproduct: %v\n", coords[0], coords[1], product)
}
